In [362]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [363]:
#loading the given datasets:
df1=pd.read_csv('python_worksheet12.csv',parse_dates=[7])
df1

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG       Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis 2016-03-09       2.5    3067.0  ...       1.0  1.0   
1             Biggin 2016-03-12       2.5    3067.0  ...       1.0  1.0   
2             Biggin 2016-04-02       2.5    3067.0  ...       1.0  0.0   
3             Rounds 2016-04-02       2.5    3067.0  ...       2.0  1.0   
4             Biggin 2017-04-03       2.5    3067.0  ...       2.0  0.0   
...              ...        ...       ...       ...  ...       ...  ...   
23673      Harcourts 2017-08-26      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart 2017-08-26      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham 2017-08-26      12.7    3085.0  ...       3.0  2.0   
23676        Village 2017-08-26       6.3    3013.0  ...       1.0  1.0   
23677        Village 2017-08-26       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

#### 1. Determine the no. of rows and no. of columns in the dataset.

In [364]:
#Shape of the data 
df1.shape

(23678, 21)

#### 2. Display the datatypes of each column.

In [365]:
df1.dtypes

Suburb                   object
Address                  object
Rooms                   float64
Type                     object
Price                   float64
Method                   object
SellerG                  object
Date             datetime64[ns]
Distance                float64
Postcode                float64
Bedroom2                float64
Bathroom                float64
Car                     float64
Landsize                float64
BuildingArea            float64
YearBuilt               float64
CouncilArea              object
Lattitude               float64
Longtitude              float64
Regionname               object
Propertycount           float64
dtype: object

#### 3. Determine the row indexes of rows having all null values (i.e. empty rows)

In [366]:
null_columns=df1.columns[df1.isnull().any()]
df1[null_columns].isnull().sum()

Suburb             131
Address            131
Rooms              131
Type               131
Price             5282
Method             131
SellerG            131
Date               131
Distance           132
Postcode           132
Bedroom2          4612
Bathroom          4615
Car               4757
Landsize          6268
BuildingArea     13660
YearBuilt        12138
CouncilArea       8022
Lattitude         4435
Longtitude        4435
Regionname         132
Propertycount      132
dtype: int64

In [367]:
print(df1[df1.isnull().all(axis=1)][null_columns])

      Suburb Address  Rooms Type  Price Method SellerG Date  Distance  \
31       NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
1145     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
1348     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
2119     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
3277     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
...      ...     ...    ...  ...    ...    ...     ...  ...       ...   
22288    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
22704    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
22918    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
23153    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   
23341    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaT       NaN   

       Postcode  ...  Bathroom  Car  Landsize  BuildingArea  YearBuilt  \
31          NaN  ...       NaN  NaN       NaN    

#### 4. Determine the total number of empty rows and their percentage in whole dataset and remove them.

In [368]:
print("Percentage of all Nan Values to the whole data set ===>", (131/23678)*100)

Percentage of all Nan Values to the whole data set ===> 0.5532561871779711


In [369]:
df=df1[null_columns].dropna(how='all')
print(df.shape)

(23547, 21)


#### 5. Determine the column names having missing values.

In [370]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

#### Determine the missing values and their percentage in each column.

In [371]:
percent_missing = df.isnull().sum() * 100 / len(df)
percent_missing

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.875398
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.004247
Postcode          0.004247
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Landsize         26.062768
BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Lattitude        18.278337
Longtitude       18.278337
Regionname        0.004247
Propertycount     0.004247
dtype: float64

#### Determine the row indexes having missing values.

In [372]:
rowswithnan = []
for index, row in df.iterrows():
    is_nan_series = row.isnull()
    if is_nan_series.any():
        rowswithnan.append(index)

print(rowswithnan)

[0, 1, 3, 5, 7, 8, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 26, 27, 28, 29, 32, 34, 35, 37, 40, 41, 42, 46, 47, 48, 49, 51, 53, 54, 55, 56, 61, 63, 65, 66, 70, 75, 76, 77, 78, 79, 80, 81, 85, 88, 90, 92, 93, 94, 95, 96, 97, 98, 100, 102, 103, 106, 108, 111, 112, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 131, 132, 133, 135, 138, 140, 141, 142, 144, 145, 147, 149, 151, 152, 154, 155, 157, 158, 159, 161, 164, 165, 166, 168, 170, 171, 172, 174, 175, 176, 180, 181, 182, 183, 185, 186, 188, 189, 190, 191, 193, 195, 197, 198, 199, 200, 201, 204, 207, 208, 209, 210, 211, 212, 213, 214, 219, 221, 222, 223, 224, 225, 226, 228, 229, 235, 238, 240, 241, 244, 246, 248, 249, 251, 253, 254, 255, 257, 258, 259, 261, 264, 266, 267, 268, 269, 271, 272, 273, 274, 275, 276, 278, 280, 281, 282, 284, 287, 288, 289, 291, 292, 294, 295, 297, 298, 299, 301, 303, 304, 305, 306, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 324, 325, 326, 327, 328, 329, 3

#### 8. Determine the missing values and their percentage in each row.

In [373]:
df.isnull().sum(axis=1) *100 / 21

0        14.285714
1         9.523810
2         0.000000
3        14.285714
4         0.000000
           ...    
23673     9.523810
23674    38.095238
23675    19.047619
23676     4.761905
23677     9.523810
Length: 23547, dtype: float64

#### 9. Determine the column names having more than 30% missing values and remove them from the dataset.

In [374]:
percent_missing= df.isnull().sum() * 100 / len(df)
percent_missing

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.875398
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.004247
Postcode          0.004247
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Landsize         26.062768
BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Lattitude        18.278337
Longtitude       18.278337
Regionname        0.004247
Propertycount     0.004247
dtype: float64

In [375]:
df.drop(['BuildingArea','YearBuilt','CouncilArea'],axis=1, inplace=True)
df.shape

(23547, 18)

#### 10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of such rows is less than 20% then remove all these rows.

In [376]:
print("Number of rows having more than 5 missing values ===>>", len(df.loc[(df.isnull().sum(axis=1)>5)]))

Number of rows having more than 5 missing values ===>> 4278


In [377]:
print("Percentage of Number of rows having more than 5 missing values ===>", (4278/23547)*100)

Percentage of Number of rows having more than 5 missing values ===> 18.16791948018856


As the Percentage of Number of rows having more than 5 missing values is less than 20% so we are going to remove 4278 rows from the data.

In [378]:
d = np.where(df.isnull().sum(axis=1)>5)
df= df.drop(df.index[d])
df.shape

(19269, 18)

#### 11. Display the description of data and percentage of missing values still left column wise.

In [379]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  19269.000000  1.508600e+04  19269.000000  19269.000000  19066.000000   
mean       2.973117  1.066695e+06     10.204365   3110.015777      2.951956   
std        0.978969  6.389331e+05      6.065786     93.730487      0.996032   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.400000e+05      6.100000   3049.000000      2.000000   
50%        3.000000  9.000000e+05      9.200000   3102.000000      3.000000   
75%        4.000000  1.320000e+06     13.000000   3150.000000      4.000000   
max       12.000000  9.000000e+06     48.100000   3977.000000     30.000000   

           Bathroom           Car       Landsize     Lattitude    Longtitude  \
count  19063.000000  18921.000000   17410.000000  19243.000000  19243.000000   
mean       1.570897      1.626235     551.783458    -37.812434    145.000287   
std        0.712684      0.974048    3544.288014      0.079926      0.106071   
min        0.000000      0.000000       0.000000    -38.184180    144.431620   
25%        1.000000      1.000000     181.000000    -37.859300    144.939300   
50%        1.000000      2.000000     448.000000    -37.809700    145.004300   
75%        2.000000      2.000000     656.000000    -37.759800    145.063100   
max       12.000000     26.000000  433014.000000    -37.407580    145.526350   

       Propertycount  
count   19269.000000  
mean     7490.249520  
std      4319.566673  
min       129.000000  
25%      4442.000000  
50%      6567.000000  
75%     10331.000000  
max     21650.000000

In [380]:
percent_missing_value= df.isnull().sum() * 100 / len(df)
percent_missing_value

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.708444
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          1.053506
Bathroom          1.069075
Car               1.806010
Landsize          9.647621
Lattitude         0.134932
Longtitude        0.134932
Regionname        0.000000
Propertycount     0.000000
dtype: float64

#### 12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column.

In [381]:
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG       Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis 2016-03-09       2.5    3067.0       2.0       1.0  1.0   
1             Biggin 2016-03-12       2.5    3067.0       2.0       1.0  1.0   
2             Biggin 2016-04-02       2.5    3067.0       2.0       1.0  0.0   
3             Rounds 2016-04-02       2.5    3067.0       3.0       2.0  1.0   
4             Biggin 2017-04-03       2.5    3067.0       3.0       2.0  0.0   
...              ...        ...       ...       ...       ...       ...  ...   
23673      Harcourts 2017-08-26      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart 2017-08-26      27.2    3024.0       NaN       NaN  NaN   
23675     Buckingham 2017-08-26      12.7    3085.0       4.0       3.0  2.0   
23676        Village 2017-08-26       6.3    3013.0       4.0       1.0  1.0   
23677        Village 2017-08-26       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

In [382]:
data_frame_list = [df]
for df in data_frame_list:
    df.bfill(inplace=True)

In [383]:
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h  1480000.0     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u  1465000.0     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u  1100000.0     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h  1100000.0      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG       Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis 2016-03-09       2.5    3067.0       2.0       1.0  1.0   
1             Biggin 2016-03-12       2.5    3067.0       2.0       1.0  1.0   
2             Biggin 2016-04-02       2.5    3067.0       2.0       1.0  0.0   
3             Rounds 2016-04-02       2.5    3067.0       3.0       2.0  1.0   
4             Biggin 2017-04-03       2.5    3067.0       3.0       2.0  0.0   
...              ...        ...       ...       ...       ...       ...  ...   
23673      Harcourts 2017-08-26      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart 2017-08-26      27.2    3024.0       4.0       3.0  2.0   
23675     Buckingham 2017-08-26      12.7    3085.0       4.0       3.0  2.0   
23676        Village 2017-08-26       6.3    3013.0       4.0       1.0  1.0   
23677        Village 2017-08-26       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674     362.0  -37.87882   144.60184   Western Metropolitan         5262.0  
23675     362.0  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

In [384]:
df.Price.isnull().sum()

0

Pandas dataframe.bfill() is used to backward fill the missing values in the dataset. It will backward fill the NaN values that are present in the above pandas dataframe. The above dataset has a pattern and I have tried to match the price backward as per the rooms and the date which is determined by the price that follows near by neighbours.

#### 13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method.

In [385]:
data_frame_list = [df]
for df in data_frame_list:
    df.ffill(inplace=True)

In [386]:
Final_percent_missing_value= df.isnull().sum() * 100 / len(df)
Final_percent_missing_value

Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64

#### 14.  Now since you have completed your data cleaning, determine the number and percentage of columns removed.

In [387]:
df.shape

(19269, 18)

In [388]:
print ("Percentage of Columns Removed from the Original Dataset" , (3/21)*100)


Percentage of Columns Removed from the Original Dataset 14.285714285714285


#### 15.  Determine the number and percentage of rows removed in whole process. Also comment briefly that according to you the amount of data lost is justified or not.

In [389]:
print ("Percentage of Rows Removed from the Original Dataset" , (4409/23678)*100)

Percentage of Rows Removed from the Original Dataset 18.620660528760872


With Respect to the above given Dataset , We have lost 18.62% of data (rows).

#### Firstly : 

We removed 131 data which had no values in all the columns hence they were of no use to us. 

#### Secondly : 

I did removed 4278 rows as the Percentage of Number of rows having more than 5 missing values that means those rows would have given us ambiguous prediction and as it creates rare labels that may lead to overfitting.

#### So, The Percentage of Removal of the 18.62% from the Original Data Set is Justified.